// @sinclair/typebox/build/esm/type/guard/type.mjs
var type_exports = {};,__export(type_exports, {
  IsAny: () => IsAny,
  IsArgument: () => IsArgument,
  IsArray: () => IsArray,
  IsAsyncIterator: () => IsAsyncIterator,
  IsBigInt: () => IsBigInt,
  IsBoolean: () => IsBoolean,
  IsComputed: () => IsComputed,
  IsConstructor: () => IsConstructor,
  IsDate: () => IsDate,
  IsFunction: () => IsFunction,
  IsImport: () => IsImport,
  IsInteger: () => IsInteger,
  IsIntersect: () => IsIntersect,
  IsIterator: () => IsIterator,
  IsKind: () => IsKind,
  IsKindOf: () => IsKindOf,
  IsLiteral: () => IsLiteral,
  IsLiteralBoolean: () => IsLiteralBoolean,
  IsLiteralNumber: () => IsLiteralNumber,
  IsLiteralString: () => IsLiteralString,
  IsLiteralValue: () => IsLiteralValue,
  IsMappedKey: () => IsMappedKey,
  IsMappedResult: () => IsMappedResult,
  IsNever: () => IsNever,
  IsNot: () => IsNot,
  IsNull: () => IsNull,
  IsNumber: () => IsNumber,
  IsObject: () => IsObject,
  IsOptional: () => IsOptional,
  IsPromise: () => IsPromise,
  IsProperties: () => IsProperties,
  IsReadonly: () => IsReadonly,
  IsRecord: () => IsRecord,
  IsRecursive: () => IsRecursive,
  IsRef: () => IsRef,
  IsRegExp: () => IsRegExp,
  IsSchema: () => IsSchema,
  IsString: () => IsString,
  IsSymbol: () => IsSymbol,
  IsTemplateLiteral: () => IsTemplateLiteral,
  IsThis: () => IsThis,
  IsTransform: () => IsTransform,
  IsTuple: () => IsTuple,
  IsUint8Array: () => IsUint8Array,
  IsUndefined: () => IsUndefined,
  IsUnion: () => IsUnion,
  IsUnionLiteral: () => IsUnionLiteral,
  IsUnknown: () => IsUnknown,
  IsUnsafe: () => IsUnsafe,
  IsVoid: () => IsVoid,
  TypeGuardUnknownTypeError: () => TypeGuardUnknownTypeError
});,var _TypeGuardUnknownTypeError = class _TypeGuardUnknownTypeError extends TypeBoxError {};,__name(_TypeGuardUnknownTypeError, "TypeGuardUnknownTypeError");,var TypeGuardUnknownTypeError = _TypeGuardUnknownTypeError,
  KnownTypes = ["Argument", "Any", "Array", "AsyncIterator", "BigInt", "Boolean", "Computed", "Constructor", "Date", "Enum", "Function", "Integer", "Intersect", "Iterator", "Literal", "MappedKey", "MappedResult", "Not", "Null", "Number", "Object", "Promise", "Record", "Ref", "RegExp", "String", "Symbol", "TemplateLiteral", "This", "Tuple", "Undefined", "Union", "Uint8Array", "Unknown", "Void"];,function IsPattern(value) {
  try {
    return new RegExp(value), !0;
  } catch {
    return !1;
  }
},__name(IsPattern, "IsPattern");,function IsControlCharacterFree(value) {
  if (!IsString(value)) return !1;
  for (let i = 0; i < value.length; i++) {
    let code = value.charCodeAt(i);
    if (code >= 7 && code <= 13 || code === 27 || code === 127) return !1;
  }
  return !0;
},__name(IsControlCharacterFree, "IsControlCharacterFree");,function IsAdditionalProperties(value) {
  return IsOptionalBoolean(value) || IsSchema(value);
},__name(IsAdditionalProperties, "IsAdditionalProperties");,function IsOptionalBigInt(value) {
  return IsUndefined(value) || IsBigInt(value);
},__name(IsOptionalBigInt, "IsOptionalBigInt");,function IsOptionalNumber(value) {
  return IsUndefined(value) || IsNumber(value);
},__name(IsOptionalNumber, "IsOptionalNumber");,function IsOptionalBoolean(value) {
  return IsUndefined(value) || IsBoolean(value);
},__name(IsOptionalBoolean, "IsOptionalBoolean");,function IsOptionalString(value) {
  return IsUndefined(value) || IsString(value);
},__name(IsOptionalString, "IsOptionalString");,function IsOptionalPattern(value) {
  return IsUndefined(value) || IsString(value) && IsControlCharacterFree(value) && IsPattern(value);
},__name(IsOptionalPattern, "IsOptionalPattern");,function IsOptionalFormat(value) {
  return IsUndefined(value) || IsString(value) && IsControlCharacterFree(value);
},__name(IsOptionalFormat, "IsOptionalFormat");,function IsOptionalSchema(value) {
  return IsUndefined(value) || IsSchema(value);
},__name(IsOptionalSchema, "IsOptionalSchema");,function IsReadonly(value) {
  return IsObject(value) && value[ReadonlyKind] === "Readonly";
},__name(IsReadonly, "IsReadonly");,function IsOptional(value) {
  return IsObject(value) && value[OptionalKind] === "Optional";
},__name(IsOptional, "IsOptional");,function IsAny(value) {
  return IsKindOf(value, "Any") && IsOptionalString(value.$id);
},__name(IsAny, "IsAny");,function IsArgument(value) {
  return IsKindOf(value, "Argument") && IsNumber(value.index);
},__name(IsArgument, "IsArgument");,function IsArray(value) {
  return IsKindOf(value, "Array") && value.type === "array" && IsOptionalString(value.$id) && IsSchema(value.items) && IsOptionalNumber(value.minItems) && IsOptionalNumber(value.maxItems) && IsOptionalBoolean(value.uniqueItems) && IsOptionalSchema(value.contains) && IsOptionalNumber(value.minContains) && IsOptionalNumber(value.maxContains);
},__name(IsArray, "IsArray");,function IsAsyncIterator(value) {
  return IsKindOf(value, "AsyncIterator") && value.type === "AsyncIterator" && IsOptionalString(value.$id) && IsSchema(value.items);
},__name(IsAsyncIterator, "IsAsyncIterator");,function IsBigInt(value) {
  return IsKindOf(value, "BigInt") && value.type === "bigint" && IsOptionalString(value.$id) && IsOptionalBigInt(value.exclusiveMaximum) && IsOptionalBigInt(value.exclusiveMinimum) && IsOptionalBigInt(value.maximum) && IsOptionalBigInt(value.minimum) && IsOptionalBigInt(value.multipleOf);
},__name(IsBigInt, "IsBigInt");,function IsBoolean(value) {
  return IsKindOf(value, "Boolean") && value.type === "boolean" && IsOptionalString(value.$id);
},__name(IsBoolean, "IsBoolean");,function IsComputed(value) {
  return IsKindOf(value, "Computed") && IsString(value.target) && IsArray(value.parameters) && value.parameters.every(schema => IsSchema(schema));
},__name(IsComputed, "IsComputed");,function IsConstructor(value) {
  return IsKindOf(value, "Constructor") && value.type === "Constructor" && IsOptionalString(value.$id) && IsArray(value.parameters) && value.parameters.every(schema => IsSchema(schema)) && IsSchema(value.returns);
},__name(IsConstructor, "IsConstructor");,function IsDate(value) {
  return IsKindOf(value, "Date") && value.type === "Date" && IsOptionalString(value.$id) && IsOptionalNumber(value.exclusiveMaximumTimestamp) && IsOptionalNumber(value.exclusiveMinimumTimestamp) && IsOptionalNumber(value.maximumTimestamp) && IsOptionalNumber(value.minimumTimestamp) && IsOptionalNumber(value.multipleOfTimestamp);
},__name(IsDate, "IsDate");,function IsFunction(value) {
  return IsKindOf(value, "Function") && value.type === "Function" && IsOptionalString(value.$id) && IsArray(value.parameters) && value.parameters.every(schema => IsSchema(schema)) && IsSchema(value.returns);
},__name(IsFunction, "IsFunction");,function IsImport(value) {
  return IsKindOf(value, "Import") && HasPropertyKey(value, "$defs") && IsObject(value.$defs) && IsProperties(value.$defs) && HasPropertyKey(value, "$ref") && IsString(value.$ref) && value.$ref in value.$defs;
},__name(IsImport, "IsImport");,function IsInteger(value) {
  return IsKindOf(value, "Integer") && value.type === "integer" && IsOptionalString(value.$id) && IsOptionalNumber(value.exclusiveMaximum) && IsOptionalNumber(value.exclusiveMinimum) && IsOptionalNumber(value.maximum) && IsOptionalNumber(value.minimum) && IsOptionalNumber(value.multipleOf);
},__name(IsInteger, "IsInteger");,function IsProperties(value) {
  return IsObject(value) && Object.entries(value).every(([key, schema]) => IsControlCharacterFree(key) && IsSchema(schema));
},__name(IsProperties, "IsProperties");,function IsIntersect(value) {
  return IsKindOf(value, "Intersect") && !(IsString(value.type) && value.type !== "object") && IsArray(value.allOf) && value.allOf.every(schema => IsSchema(schema) && !IsTransform(schema)) && IsOptionalString(value.type) && (IsOptionalBoolean(value.unevaluatedProperties) || IsOptionalSchema(value.unevaluatedProperties)) && IsOptionalString(value.$id);
},__name(IsIntersect, "IsIntersect");,function IsIterator(value) {
  return IsKindOf(value, "Iterator") && value.type === "Iterator" && IsOptionalString(value.$id) && IsSchema(value.items);
},__name(IsIterator, "IsIterator");,function IsKindOf(value, kind) {
  return IsObject(value) && Kind in value && value[Kind] === kind;
},__name(IsKindOf, "IsKindOf");,function IsLiteralString(value) {
  return IsLiteral(value) && IsString(value.const);
},__name(IsLiteralString, "IsLiteralString");,function IsLiteralNumber(value) {
  return IsLiteral(value) && IsNumber(value.const);
},__name(IsLiteralNumber, "IsLiteralNumber");,function IsLiteralBoolean(value) {
  return IsLiteral(value) && IsBoolean(value.const);
},__name(IsLiteralBoolean, "IsLiteralBoolean");,function IsLiteral(value) {
  return IsKindOf(value, "Literal") && IsOptionalString(value.$id) && IsLiteralValue(value.const);
},__name(IsLiteral, "IsLiteral");,function IsLiteralValue(value) {
  return IsBoolean(value) || IsNumber(value) || IsString(value);
},__name(IsLiteralValue, "IsLiteralValue");,function IsMappedKey(value) {
  return IsKindOf(value, "MappedKey") && IsArray(value.keys) && value.keys.every(key => IsNumber(key) || IsString(key));
},__name(IsMappedKey, "IsMappedKey");,function IsMappedResult(value) {
  return IsKindOf(value, "MappedResult") && IsProperties(value.properties);
},__name(IsMappedResult, "IsMappedResult");,function IsNever(value) {
  return IsKindOf(value, "Never") && IsObject(value.not) && Object.getOwnPropertyNames(value.not).length === 0;
},__name(IsNever, "IsNever");,function IsNot(value) {
  return IsKindOf(value, "Not") && IsSchema(value.not);
},__name(IsNot, "IsNot");,function IsNull(value) {
  return IsKindOf(value, "Null") && value.type === "null" && IsOptionalString(value.$id);
},__name(IsNull, "IsNull");,function IsNumber(value) {
  return IsKindOf(value, "Number") && value.type === "number" && IsOptionalString(value.$id) && IsOptionalNumber(value.exclusiveMaximum) && IsOptionalNumber(value.exclusiveMinimum) && IsOptionalNumber(value.maximum) && IsOptionalNumber(value.minimum) && IsOptionalNumber(value.multipleOf);
},__name(IsNumber, "IsNumber");,function IsObject(value) {
  return IsKindOf(value, "Object") && value.type === "object" && IsOptionalString(value.$id) && IsProperties(value.properties) && IsAdditionalProperties(value.additionalProperties) && IsOptionalNumber(value.minProperties) && IsOptionalNumber(value.maxProperties);
},__name(IsObject, "IsObject");,function IsPromise(value) {
  return IsKindOf(value, "Promise") && value.type === "Promise" && IsOptionalString(value.$id) && IsSchema(value.item);
},__name(IsPromise, "IsPromise");,function IsRecord(value) {
  return IsKindOf(value, "Record") && value.type === "object" && IsOptionalString(value.$id) && IsAdditionalProperties(value.additionalProperties) && IsObject(value.patternProperties) && (schema => {
    let keys = Object.getOwnPropertyNames(schema.patternProperties);
    return keys.length === 1 && IsPattern(keys[0]) && IsObject(schema.patternProperties) && IsSchema(schema.patternProperties[keys[0]]);
  })(value);
},__name(IsRecord, "IsRecord");,function IsRecursive(value) {
  return IsObject(value) && Hint in value && value[Hint] === "Recursive";
},__name(IsRecursive, "IsRecursive");,function IsRef(value) {
  return IsKindOf(value, "Ref") && IsOptionalString(value.$id) && IsString(value.$ref);
},__name(IsRef, "IsRef");,function IsRegExp(value) {
  return IsKindOf(value, "RegExp") && IsOptionalString(value.$id) && IsString(value.source) && IsString(value.flags) && IsOptionalNumber(value.maxLength) && IsOptionalNumber(value.minLength);
},__name(IsRegExp, "IsRegExp");,function IsString(value) {
  return IsKindOf(value, "String") && value.type === "string" && IsOptionalString(value.$id) && IsOptionalNumber(value.minLength) && IsOptionalNumber(value.maxLength) && IsOptionalPattern(value.pattern) && IsOptionalFormat(value.format);
},__name(IsString, "IsString");,function IsSymbol(value) {
  return IsKindOf(value, "Symbol") && value.type === "symbol" && IsOptionalString(value.$id);
},__name(IsSymbol, "IsSymbol");,function IsTemplateLiteral(value) {
  return IsKindOf(value, "TemplateLiteral") && value.type === "string" && IsString(value.pattern) && value.pattern[0] === "^" && value.pattern[value.pattern.length - 1] === "$";
},__name(IsTemplateLiteral, "IsTemplateLiteral");,function IsThis(value) {
  return IsKindOf(value, "This") && IsOptionalString(value.$id) && IsString(value.$ref);
},__name(IsThis, "IsThis");,function IsTransform(value) {
  return IsObject(value) && TransformKind in value;
},__name(IsTransform, "IsTransform");,function IsTuple(value) {
  return IsKindOf(value, "Tuple") && value.type === "array" && IsOptionalString(value.$id) && IsNumber(value.minItems) && IsNumber(value.maxItems) && value.minItems === value.maxItems && (IsUndefined(value.items) && IsUndefined(value.additionalItems) && value.minItems === 0 || IsArray(value.items) && value.items.every(schema => IsSchema(schema)));
},__name(IsTuple, "IsTuple");,function IsUndefined(value) {
  return IsKindOf(value, "Undefined") && value.type === "undefined" && IsOptionalString(value.$id);
},__name(IsUndefined, "IsUndefined");,function IsUnionLiteral(value) {
  return IsUnion(value) && value.anyOf.every(schema => IsLiteralString(schema) || IsLiteralNumber(schema));
},__name(IsUnionLiteral, "IsUnionLiteral");,function IsUnion(value) {
  return IsKindOf(value, "Union") && IsOptionalString(value.$id) && IsObject(value) && IsArray(value.anyOf) && value.anyOf.every(schema => IsSchema(schema));
},__name(IsUnion, "IsUnion");,function IsUint8Array(value) {
  return IsKindOf(value, "Uint8Array") && value.type === "Uint8Array" && IsOptionalString(value.$id) && IsOptionalNumber(value.minByteLength) && IsOptionalNumber(value.maxByteLength);
},__name(IsUint8Array, "IsUint8Array");,function IsUnknown(value) {
  return IsKindOf(value, "Unknown") && IsOptionalString(value.$id);
},__name(IsUnknown, "IsUnknown");,function IsUnsafe(value) {
  return IsKindOf(value, "Unsafe");
},__name(IsUnsafe, "IsUnsafe");,function IsVoid(value) {
  return IsKindOf(value, "Void") && value.type === "void" && IsOptionalString(value.$id);
},__name(IsVoid, "IsVoid");,function IsKind(value) {
  return IsObject(value) && Kind in value && IsString(value[Kind]) && !KnownTypes.includes(value[Kind]);
},__name(IsKind, "IsKind");,function IsSchema(value) {
  return IsObject(value) && (IsAny(value) || IsArgument(value) || IsArray(value) || IsBoolean(value) || IsBigInt(value) || IsAsyncIterator(value) || IsComputed(value) || IsConstructor(value) || IsDate(value) || IsFunction(value) || IsInteger(value) || IsIntersect(value) || IsIterator(value) || IsLiteral(value) || IsMappedKey(value) || IsMappedResult(value) || IsNever(value) || IsNot(value) || IsNull(value) || IsNumber(value) || IsObject(value) || IsPromise(value) || IsRecord(value) || IsRef(value) || IsRegExp(value) || IsString(value) || IsSymbol(value) || IsTemplateLiteral(value) || IsThis(value) || IsTuple(value) || IsUndefined(value) || IsUnion(value) || IsUint8Array(value) || IsUnknown(value) || IsUnsafe(value) || IsVoid(value) || IsKind(value));
},__name(IsSchema, "IsSchema");
