// @sinclair/typebox/build/esm/type/guard/value.mjs
var value_exports = {};,__export(value_exports, {
  HasPropertyKey: () => HasPropertyKey,
  IsArray: () => IsArray,
  IsAsyncIterator: () => IsAsyncIterator,
  IsBigInt: () => IsBigInt,
  IsBoolean: () => IsBoolean,
  IsDate: () => IsDate,
  IsFunction: () => IsFunction,
  IsIterator: () => IsIterator,
  IsNull: () => IsNull,
  IsNumber: () => IsNumber,
  IsObject: () => IsObject,
  IsRegExp: () => IsRegExp,
  IsString: () => IsString,
  IsSymbol: () => IsSymbol,
  IsUint8Array: () => IsUint8Array,
  IsUndefined: () => IsUndefined
});,function HasPropertyKey(value, key) {
  return key in value;
},__name(HasPropertyKey, "HasPropertyKey");,function IsAsyncIterator(value) {
  return IsObject(value) && !IsArray(value) && !IsUint8Array(value) && Symbol.asyncIterator in value;
},__name(IsAsyncIterator, "IsAsyncIterator");,function IsArray(value) {
  return Array.isArray(value);
},__name(IsArray, "IsArray");,function IsBigInt(value) {
  return typeof value == "bigint";
},__name(IsBigInt, "IsBigInt");,function IsBoolean(value) {
  return typeof value == "boolean";
},__name(IsBoolean, "IsBoolean");,function IsDate(value) {
  return value instanceof globalThis.Date;
},__name(IsDate, "IsDate");,function IsFunction(value) {
  return typeof value == "function";
},__name(IsFunction, "IsFunction");,function IsIterator(value) {
  return IsObject(value) && !IsArray(value) && !IsUint8Array(value) && Symbol.iterator in value;
},__name(IsIterator, "IsIterator");,function IsNull(value) {
  return value === null;
},__name(IsNull, "IsNull");,function IsNumber(value) {
  return typeof value == "number";
},__name(IsNumber, "IsNumber");,function IsObject(value) {
  return typeof value == "object" && value !== null;
},__name(IsObject, "IsObject");,function IsRegExp(value) {
  return value instanceof globalThis.RegExp;
},__name(IsRegExp, "IsRegExp");,function IsString(value) {
  return typeof value == "string";
},__name(IsString, "IsString");,function IsSymbol(value) {
  return typeof value == "symbol";
},__name(IsSymbol, "IsSymbol");,function IsUint8Array(value) {
  return value instanceof globalThis.Uint8Array;
},__name(IsUint8Array, "IsUint8Array");,function IsUndefined(value) {
  return value === void 0;
},__name(IsUndefined, "IsUndefined");
