// parse-path/lib/index.js
var require_lib = __commonJSMin((exports, module) => {
  "use strict";

  var protocols = JW();
  function parsePath(url) {
    var output = {
      protocols: [],
      protocol: null,
      port: null,
      resource: "",
      host: "",
      user: "",
      password: "",
      pathname: "",
      hash: "",
      search: "",
      href: url,
      query: {},
      parse_failed: !1
    };
    try {
      var parsed = new URL(url);
      output.protocols = protocols(parsed), output.protocol = output.protocols[0], output.port = parsed.port, output.resource = parsed.hostname, output.host = parsed.host, output.user = parsed.username || "", output.password = parsed.password || "", output.pathname = parsed.pathname, output.hash = parsed.hash.slice(1), output.search = parsed.search.slice(1), output.href = parsed.href, output.query = Object.fromEntries(parsed.searchParams);
    } catch {
      output.protocols = ["file"], output.protocol = output.protocols[0], output.port = "", output.resource = "", output.user = "", output.pathname = "", output.hash = "", output.search = "", output.href = url, output.query = {}, output.parse_failed = !0;
    }
    return output;
  }
  __name(parsePath, "parsePath");
  module.exports = parsePath;
});
