// shiki/dist/langs/angular-html.mjs
var angular_html_exports = {};,__export(angular_html_exports, {
  default: () => angular_html
});,var lang,
  angular_html,
  init_angular_html = __esmMin(() => {
    init_html();
    init_angular_expression();
    init_angular_let_declaration();
    init_angular_template();
    init_angular_template_blocks();
    lang = Object.freeze({
      displayName: "Angular HTML",
      injections: {
        "R:text.html - (comment.block, text.html meta.embedded, meta.tag.*.*.html, meta.tag.*.*.*.html, meta.tag.*.*.*.*.html)": {
          comment: "Uses R: to ensure this matches after any other injections.",
          patterns: [{
            match: "<",
            name: "invalid.illegal.bad-angle-bracket.html"
          }]
        }
      },
      name: "angular-html",
      patterns: [{
        include: "text.html.basic#core-minus-invalid"
      }, {
        begin: "(</?)(\\w[^\\s>]*)(?<!/)",
        beginCaptures: {
          1: {
            name: "punctuation.definition.tag.begin.html"
          },
          2: {
            name: "entity.name.tag.html"
          }
        },
        end: "((?: ?/)?>)",
        endCaptures: {
          1: {
            name: "punctuation.definition.tag.end.html"
          }
        },
        name: "meta.tag.other.unrecognized.html.derivative",
        patterns: [{
          include: "text.html.basic#attribute"
        }]
      }],
      scopeName: "text.html.derivative.ng",
      embeddedLangs: ["html", "angular-expression", "angular-let-declaration", "angular-template", "angular-template-blocks"]
    }), angular_html = [...html, ...angular_expression, ...angular_let_declaration, ...angular_template, ...angular_template_blocks, lang];
  });
