// undici/lib/web/fetch/dispatcher-weakref.js
var require_dispatcher_weakref = __commonJSMin((exports, module) => {
  "use strict";

  var {
      kConnected: kConnected,
      kSize: kSize
    } = fo(),
    _CompatWeakRef = class _CompatWeakRef {
      constructor(value) {
        this.value = value;
      }
      deref() {
        return this.value[kConnected] === 0 && this.value[kSize] === 0 ? void 0 : this.value;
      }
    };
  __name(_CompatWeakRef, "CompatWeakRef");
  var CompatWeakRef = _CompatWeakRef,
    _CompatFinalizer = class _CompatFinalizer {
      constructor(finalizer) {
        this.finalizer = finalizer;
      }
      register(dispatcher, key) {
        dispatcher.on && dispatcher.on("disconnect", () => {
          dispatcher[kConnected] === 0 && dispatcher[kSize] === 0 && this.finalizer(key);
        });
      }
      unregister(key) {}
    };
  __name(_CompatFinalizer, "CompatFinalizer");
  var CompatFinalizer = _CompatFinalizer;
  module.exports = function () {
    return process.env.NODE_V8_COVERAGE && process.version.startsWith("v18") ? (process._rawDebug("Using compatibility WeakRef and FinalizationRegistry"), {
      WeakRef: CompatWeakRef,
      FinalizationRegistry: CompatFinalizer
    }) : {
      WeakRef,
      FinalizationRegistry
    };
  };
});
