// vscode-jsonrpc/lib/common/api.js
var require_api = __commonJSMin(exports => {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: !0
  });
  exports.ProgressType = exports.ProgressToken = exports.createMessageConnection = exports.NullLogger = exports.ConnectionOptions = exports.ConnectionStrategy = exports.AbstractMessageBuffer = exports.WriteableStreamMessageWriter = exports.AbstractMessageWriter = exports.MessageWriter = exports.ReadableStreamMessageReader = exports.AbstractMessageReader = exports.MessageReader = exports.SharedArrayReceiverStrategy = exports.SharedArraySenderStrategy = exports.CancellationToken = exports.CancellationTokenSource = exports.Emitter = exports.Event = exports.Disposable = exports.LRUCache = exports.Touch = exports.LinkedMap = exports.ParameterStructures = exports.NotificationType9 = exports.NotificationType8 = exports.NotificationType7 = exports.NotificationType6 = exports.NotificationType5 = exports.NotificationType4 = exports.NotificationType3 = exports.NotificationType2 = exports.NotificationType1 = exports.NotificationType0 = exports.NotificationType = exports.ErrorCodes = exports.ResponseError = exports.RequestType9 = exports.RequestType8 = exports.RequestType7 = exports.RequestType6 = exports.RequestType5 = exports.RequestType4 = exports.RequestType3 = exports.RequestType2 = exports.RequestType1 = exports.RequestType0 = exports.RequestType = exports.Message = exports.RAL = void 0;
  exports.MessageStrategy = exports.CancellationStrategy = exports.CancellationSenderStrategy = exports.CancellationReceiverStrategy = exports.ConnectionError = exports.ConnectionErrors = exports.LogTraceNotification = exports.SetTraceNotification = exports.TraceFormat = exports.TraceValues = exports.Trace = void 0;
  var messages_1 = XJ();
  Object.defineProperty(exports, "Message", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.Message;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType0", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType0;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType1", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType1;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType2", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType2;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType3", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType3;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType4", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType4;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType5", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType5;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType6", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType6;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType7", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType7;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType8", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType8;
    }, "get")
  });
  Object.defineProperty(exports, "RequestType9", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.RequestType9;
    }, "get")
  });
  Object.defineProperty(exports, "ResponseError", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.ResponseError;
    }, "get")
  });
  Object.defineProperty(exports, "ErrorCodes", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.ErrorCodes;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType0", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType0;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType1", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType1;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType2", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType2;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType3", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType3;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType4", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType4;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType5", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType5;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType6", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType6;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType7", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType7;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType8", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType8;
    }, "get")
  });
  Object.defineProperty(exports, "NotificationType9", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.NotificationType9;
    }, "get")
  });
  Object.defineProperty(exports, "ParameterStructures", {
    enumerable: !0,
    get: __name(function () {
      return messages_1.ParameterStructures;
    }, "get")
  });
  var linkedMap_1 = aV();
  Object.defineProperty(exports, "LinkedMap", {
    enumerable: !0,
    get: __name(function () {
      return linkedMap_1.LinkedMap;
    }, "get")
  });
  Object.defineProperty(exports, "LRUCache", {
    enumerable: !0,
    get: __name(function () {
      return linkedMap_1.LRUCache;
    }, "get")
  });
  Object.defineProperty(exports, "Touch", {
    enumerable: !0,
    get: __name(function () {
      return linkedMap_1.Touch;
    }, "get")
  });
  var disposable_1 = fQe();
  Object.defineProperty(exports, "Disposable", {
    enumerable: !0,
    get: __name(function () {
      return disposable_1.Disposable;
    }, "get")
  });
  var events_1 = N1();
  Object.defineProperty(exports, "Event", {
    enumerable: !0,
    get: __name(function () {
      return events_1.Event;
    }, "get")
  });
  Object.defineProperty(exports, "Emitter", {
    enumerable: !0,
    get: __name(function () {
      return events_1.Emitter;
    }, "get")
  });
  var cancellation_1 = bL();
  Object.defineProperty(exports, "CancellationTokenSource", {
    enumerable: !0,
    get: __name(function () {
      return cancellation_1.CancellationTokenSource;
    }, "get")
  });
  Object.defineProperty(exports, "CancellationToken", {
    enumerable: !0,
    get: __name(function () {
      return cancellation_1.CancellationToken;
    }, "get")
  });
  var sharedArrayCancellation_1 = bQe();
  Object.defineProperty(exports, "SharedArraySenderStrategy", {
    enumerable: !0,
    get: __name(function () {
      return sharedArrayCancellation_1.SharedArraySenderStrategy;
    }, "get")
  });
  Object.defineProperty(exports, "SharedArrayReceiverStrategy", {
    enumerable: !0,
    get: __name(function () {
      return sharedArrayCancellation_1.SharedArrayReceiverStrategy;
    }, "get")
  });
  var messageReader_1 = wQe();
  Object.defineProperty(exports, "MessageReader", {
    enumerable: !0,
    get: __name(function () {
      return messageReader_1.MessageReader;
    }, "get")
  });
  Object.defineProperty(exports, "AbstractMessageReader", {
    enumerable: !0,
    get: __name(function () {
      return messageReader_1.AbstractMessageReader;
    }, "get")
  });
  Object.defineProperty(exports, "ReadableStreamMessageReader", {
    enumerable: !0,
    get: __name(function () {
      return messageReader_1.ReadableStreamMessageReader;
    }, "get")
  });
  var messageWriter_1 = EQe();
  Object.defineProperty(exports, "MessageWriter", {
    enumerable: !0,
    get: __name(function () {
      return messageWriter_1.MessageWriter;
    }, "get")
  });
  Object.defineProperty(exports, "AbstractMessageWriter", {
    enumerable: !0,
    get: __name(function () {
      return messageWriter_1.AbstractMessageWriter;
    }, "get")
  });
  Object.defineProperty(exports, "WriteableStreamMessageWriter", {
    enumerable: !0,
    get: __name(function () {
      return messageWriter_1.WriteableStreamMessageWriter;
    }, "get")
  });
  var messageBuffer_1 = BQe();
  Object.defineProperty(exports, "AbstractMessageBuffer", {
    enumerable: !0,
    get: __name(function () {
      return messageBuffer_1.AbstractMessageBuffer;
    }, "get")
  });
  var connection_1 = SQe();
  Object.defineProperty(exports, "ConnectionStrategy", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.ConnectionStrategy;
    }, "get")
  });
  Object.defineProperty(exports, "ConnectionOptions", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.ConnectionOptions;
    }, "get")
  });
  Object.defineProperty(exports, "NullLogger", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.NullLogger;
    }, "get")
  });
  Object.defineProperty(exports, "createMessageConnection", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.createMessageConnection;
    }, "get")
  });
  Object.defineProperty(exports, "ProgressToken", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.ProgressToken;
    }, "get")
  });
  Object.defineProperty(exports, "ProgressType", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.ProgressType;
    }, "get")
  });
  Object.defineProperty(exports, "Trace", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.Trace;
    }, "get")
  });
  Object.defineProperty(exports, "TraceValues", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.TraceValues;
    }, "get")
  });
  Object.defineProperty(exports, "TraceFormat", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.TraceFormat;
    }, "get")
  });
  Object.defineProperty(exports, "SetTraceNotification", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.SetTraceNotification;
    }, "get")
  });
  Object.defineProperty(exports, "LogTraceNotification", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.LogTraceNotification;
    }, "get")
  });
  Object.defineProperty(exports, "ConnectionErrors", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.ConnectionErrors;
    }, "get")
  });
  Object.defineProperty(exports, "ConnectionError", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.ConnectionError;
    }, "get")
  });
  Object.defineProperty(exports, "CancellationReceiverStrategy", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.CancellationReceiverStrategy;
    }, "get")
  });
  Object.defineProperty(exports, "CancellationSenderStrategy", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.CancellationSenderStrategy;
    }, "get")
  });
  Object.defineProperty(exports, "CancellationStrategy", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.CancellationStrategy;
    }, "get")
  });
  Object.defineProperty(exports, "MessageStrategy", {
    enumerable: !0,
    get: __name(function () {
      return connection_1.MessageStrategy;
    }, "get")
  });
  var ral_1 = zy();
  exports.RAL = ral_1.default;
});
