// vscode-languageclient/lib/common/hover.js
var require_hover = __commonJSMin(exports => {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: !0
  });
  exports.HoverFeature = void 0;
  var vscode_1 = require("vscode"),
    vscode_languageserver_protocol_1 = Un(),
    features_1 = Oi(),
    UUID = yo(),
    _HoverFeature = class _HoverFeature extends features_1.TextDocumentLanguageFeature {
      constructor(client) {
        super(client, vscode_languageserver_protocol_1.HoverRequest.type);
      }
      fillClientCapabilities(capabilities) {
        let hoverCapability = (0, features_1.ensure)((0, features_1.ensure)(capabilities, "textDocument"), "hover");
        hoverCapability.dynamicRegistration = !0, hoverCapability.contentFormat = [vscode_languageserver_protocol_1.MarkupKind.Markdown, vscode_languageserver_protocol_1.MarkupKind.PlainText];
      }
      initialize(capabilities, documentSelector) {
        let options = this.getRegistrationOptions(documentSelector, capabilities.hoverProvider);
        options && this.register({
          id: UUID.generateUuid(),
          registerOptions: options
        });
      }
      registerLanguageProvider(options) {
        let selector = options.documentSelector,
          provider = {
            provideHover: __name((document, position, token) => {
              let client = this._client,
                provideHover = __name((document, position, token) => client.sendRequest(vscode_languageserver_protocol_1.HoverRequest.type, client.code2ProtocolConverter.asTextDocumentPositionParams(document, position), token).then(result => token.isCancellationRequested ? null : client.protocol2CodeConverter.asHover(result), error => client.handleFailedRequest(vscode_languageserver_protocol_1.HoverRequest.type, token, error, null)), "provideHover"),
                middleware = client.middleware;
              return middleware.provideHover ? middleware.provideHover(document, position, token, provideHover) : provideHover(document, position, token);
            }, "provideHover")
          };
        return [this.registerProvider(selector, provider), provider];
      }
      registerProvider(selector, provider) {
        return vscode_1.languages.registerHoverProvider(this._client.protocol2CodeConverter.asDocumentSelector(selector), provider);
      }
    };
  __name(_HoverFeature, "HoverFeature");
  var HoverFeature = _HoverFeature;
  exports.HoverFeature = HoverFeature;
});
