// vscode-languageclient/lib/common/reference.js
var require_reference = __commonJSMin(exports => {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: !0
  });
  exports.ReferencesFeature = void 0;
  var vscode_1 = require("vscode"),
    vscode_languageserver_protocol_1 = Un(),
    features_1 = Oi(),
    UUID = yo(),
    _ReferencesFeature = class _ReferencesFeature extends features_1.TextDocumentLanguageFeature {
      constructor(client) {
        super(client, vscode_languageserver_protocol_1.ReferencesRequest.type);
      }
      fillClientCapabilities(capabilities) {
        (0, features_1.ensure)((0, features_1.ensure)(capabilities, "textDocument"), "references").dynamicRegistration = !0;
      }
      initialize(capabilities, documentSelector) {
        let options = this.getRegistrationOptions(documentSelector, capabilities.referencesProvider);
        options && this.register({
          id: UUID.generateUuid(),
          registerOptions: options
        });
      }
      registerLanguageProvider(options) {
        let selector = options.documentSelector,
          provider = {
            provideReferences: __name((document, position, options, token) => {
              let client = this._client,
                _providerReferences = __name((document, position, options, token) => client.sendRequest(vscode_languageserver_protocol_1.ReferencesRequest.type, client.code2ProtocolConverter.asReferenceParams(document, position, options), token).then(result => token.isCancellationRequested ? null : client.protocol2CodeConverter.asReferences(result, token), error => client.handleFailedRequest(vscode_languageserver_protocol_1.ReferencesRequest.type, token, error, null)), "_providerReferences"),
                middleware = client.middleware;
              return middleware.provideReferences ? middleware.provideReferences(document, position, options, token, _providerReferences) : _providerReferences(document, position, options, token);
            }, "provideReferences")
          };
        return [this.registerProvider(selector, provider), provider];
      }
      registerProvider(selector, provider) {
        return vscode_1.languages.registerReferenceProvider(this._client.protocol2CodeConverter.asDocumentSelector(selector), provider);
      }
    };
  __name(_ReferencesFeature, "ReferencesFeature");
  var ReferencesFeature = _ReferencesFeature;
  exports.ReferencesFeature = ReferencesFeature;
});
