// source-map/lib/mapping-list.js
var require_mapping_list = __commonJSMin(exports => {
  var util = yB();
  function generatedPositionAfter(mappingA, mappingB) {
    var lineA = mappingA.generatedLine,
      lineB = mappingB.generatedLine,
      columnA = mappingA.generatedColumn,
      columnB = mappingB.generatedColumn;
    return lineB > lineA || lineB == lineA && columnB >= columnA || util.compareByGeneratedPositionsInflated(mappingA, mappingB) <= 0;
  }
  __name(generatedPositionAfter, "generatedPositionAfter");
  function MappingList() {
    this._array = [], this._sorted = !0, this._last = {
      generatedLine: -1,
      generatedColumn: 0
    };
  }
  __name(MappingList, "MappingList");
  MappingList.prototype.unsortedForEach = __name(function (aCallback, aThisArg) {
    this._array.forEach(aCallback, aThisArg);
  }, "MappingList_forEach");
  MappingList.prototype.add = __name(function (aMapping) {
    generatedPositionAfter(this._last, aMapping) ? (this._last = aMapping, this._array.push(aMapping)) : (this._sorted = !1, this._array.push(aMapping));
  }, "MappingList_add");
  MappingList.prototype.toArray = __name(function () {
    return this._sorted || (this._array.sort(util.compareByGeneratedPositionsInflated), this._sorted = !0), this._array;
  }, "MappingList_toArray");
  exports.MappingList = MappingList;
});
