// undici/lib/dispatcher/fixed-queue.js
var require_fixed_queue = __commonJSMin((exports, module) => {
  "use strict";

  var _FixedCircularBuffer = class _FixedCircularBuffer {
    constructor() {
      this.bottom = 0, this.top = 0, this.list = new Array(2048), this.next = null;
    }
    isEmpty() {
      return this.top === this.bottom;
    }
    isFull() {
      return (this.top + 1 & 2047) === this.bottom;
    }
    push(data) {
      this.list[this.top] = data, this.top = this.top + 1 & 2047;
    }
    shift() {
      let nextItem = this.list[this.bottom];
      return nextItem === void 0 ? null : (this.list[this.bottom] = void 0, this.bottom = this.bottom + 1 & 2047, nextItem);
    }
  };
  __name(_FixedCircularBuffer, "FixedCircularBuffer");
  var FixedCircularBuffer = _FixedCircularBuffer,
    _a;
  module.exports = (_a = class {
    constructor() {
      this.head = this.tail = new FixedCircularBuffer();
    }
    isEmpty() {
      return this.head.isEmpty();
    }
    push(data) {
      this.head.isFull() && (this.head = this.head.next = new FixedCircularBuffer()), this.head.push(data);
    }
    shift() {
      let tail = this.tail,
        next = tail.shift();
      return tail.isEmpty() && tail.next !== null && (this.tail = tail.next), next;
    }
  }, __name(_a, "FixedQueue"), _a);
});
