// vscode-languageserver-protocol/lib/common/messages.js
var require_messages = __commonJSMin(exports => {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: !0
  });
  exports.ProtocolNotificationType = exports.ProtocolNotificationType0 = exports.ProtocolRequestType = exports.ProtocolRequestType0 = exports.RegistrationType = exports.MessageDirection = void 0;
  var vscode_jsonrpc_1 = U_(),
    MessageDirection;
  (function (MessageDirection) {
    MessageDirection.clientToServer = "clientToServer", MessageDirection.serverToClient = "serverToClient", MessageDirection.both = "both";
  })(MessageDirection || (exports.MessageDirection = MessageDirection = {}));
  var _RegistrationType = class _RegistrationType {
    constructor(method) {
      this.method = method;
    }
  };
  __name(_RegistrationType, "RegistrationType");
  var RegistrationType = _RegistrationType;
  exports.RegistrationType = RegistrationType;
  var _ProtocolRequestType0 = class _ProtocolRequestType0 extends vscode_jsonrpc_1.RequestType0 {
    constructor(method) {
      super(method);
    }
  };
  __name(_ProtocolRequestType0, "ProtocolRequestType0");
  var ProtocolRequestType0 = _ProtocolRequestType0;
  exports.ProtocolRequestType0 = ProtocolRequestType0;
  var _ProtocolRequestType = class _ProtocolRequestType extends vscode_jsonrpc_1.RequestType {
    constructor(method) {
      super(method, vscode_jsonrpc_1.ParameterStructures.byName);
    }
  };
  __name(_ProtocolRequestType, "ProtocolRequestType");
  var ProtocolRequestType = _ProtocolRequestType;
  exports.ProtocolRequestType = ProtocolRequestType;
  var _ProtocolNotificationType0 = class _ProtocolNotificationType0 extends vscode_jsonrpc_1.NotificationType0 {
    constructor(method) {
      super(method);
    }
  };
  __name(_ProtocolNotificationType0, "ProtocolNotificationType0");
  var ProtocolNotificationType0 = _ProtocolNotificationType0;
  exports.ProtocolNotificationType0 = ProtocolNotificationType0;
  var _ProtocolNotificationType = class _ProtocolNotificationType extends vscode_jsonrpc_1.NotificationType {
    constructor(method) {
      super(method, vscode_jsonrpc_1.ParameterStructures.byName);
    }
  };
  __name(_ProtocolNotificationType, "ProtocolNotificationType");
  var ProtocolNotificationType = _ProtocolNotificationType;
  exports.ProtocolNotificationType = ProtocolNotificationType;
});
