// @microsoft/tiktokenizer/dist/index.js
var require_dist = __commonJSMin(exports => {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: !0
  });
  exports.createTokenizer = exports.createByEncoderName = exports.createByModelName = exports.getSpecialTokensByModel = exports.getSpecialTokensByEncoder = exports.getRegexByModel = exports.getRegexByEncoder = exports.MODEL_TO_ENCODING = exports.TikTokenizer = void 0;
  var tikTokenizer_1 = hY();
  Object.defineProperty(exports, "TikTokenizer", {
    enumerable: !0,
    get: __name(function () {
      return tikTokenizer_1.TikTokenizer;
    }, "get")
  });
  var tokenizerBuilder_1 = eIe();
  Object.defineProperty(exports, "MODEL_TO_ENCODING", {
    enumerable: !0,
    get: __name(function () {
      return tokenizerBuilder_1.MODEL_TO_ENCODING;
    }, "get")
  });
  Object.defineProperty(exports, "getRegexByEncoder", {
    enumerable: !0,
    get: __name(function () {
      return tokenizerBuilder_1.getRegexByEncoder;
    }, "get")
  });
  Object.defineProperty(exports, "getRegexByModel", {
    enumerable: !0,
    get: __name(function () {
      return tokenizerBuilder_1.getRegexByModel;
    }, "get")
  });
  Object.defineProperty(exports, "getSpecialTokensByEncoder", {
    enumerable: !0,
    get: __name(function () {
      return tokenizerBuilder_1.getSpecialTokensByEncoder;
    }, "get")
  });
  Object.defineProperty(exports, "getSpecialTokensByModel", {
    enumerable: !0,
    get: __name(function () {
      return tokenizerBuilder_1.getSpecialTokensByModel;
    }, "get")
  });
  Object.defineProperty(exports, "createByModelName", {
    enumerable: !0,
    get: __name(function () {
      return tokenizerBuilder_1.createByModelName;
    }, "get")
  });
  Object.defineProperty(exports, "createByEncoderName", {
    enumerable: !0,
    get: __name(function () {
      return tokenizerBuilder_1.createByEncoderName;
    }, "get")
  });
  Object.defineProperty(exports, "createTokenizer", {
    enumerable: !0,
    get: __name(function () {
      return tokenizerBuilder_1.createTokenizer;
    }, "get")
  });
});
